src/link/link_info/iptunnel.rs

```diff
@@ -0,0 +1,391 @@
+const IFLA_IPTUN_6RD_PREFIX: u16 = 11;
+const IFLA_IPTUN_6RD_RELAY_PREFIX: u16 = 12;
+const IFLA_IPTUN_6RD_PREFIXLEN: u16 = 13;
+const IFLA_IPTUN_6RD_RELAY_PREFIXLEN: u16 = 14;
+const IFLA_IPTUN_ENCAP_TYPE: u16 = 15;
+const IFLA_IPTUN_ENCAP_FLAGS: u16 = 16;
+const IFLA_IPTUN_ENCAP_SPORT: u16 = 17;
+const IFLA_IPTUN_ENCAP_DPORT: u16 = 18;
+const IFLA_IPTUN_COLLECT_METADATA: u16 = 19;
+const IFLA_IPTUN_FWMARK: u16 = 20;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub enum InfoIpTunnel {
+    Link(u32),
+    Local(IpAddr),
+    Remote(IpAddr),
+    Ttl(u8),
+    Tos(u8),
+    EncapLimit(u8),
+    FlowInfo(u32),
+    Flags(TunnelFlags),
+    Protocol(IpProtocol),
+    PMtuDisc(bool),
+    Ipv6RdPrefix(Ipv6Addr),
+    Ipv6RdRelayPrefix(Ipv6Addr),
+    Ipv6RdPrefixLen(u16),
+    Ipv6RdRelayPrefixLen(u16),
+    EncapType(TunnelEncapType),
+    EncapFlags(TunnelEncapFlags),
+    EncapSPort(u16),
+    EncapDPort(u16),
+    CollectMetada(bool),
+    FwMark(u32),
+    Other(DefaultNla),
+}
+
+impl Nla for InfoIpTunnel {
+    fn value_len(&self) -> usize {
+        use self::InfoIpTunnel::*;
+        match self {
+            Ipv6RdPrefix(_) | Ipv6RdRelayPrefix(_) => 16,
+            Local(value) | Remote(value) => match value {
+                IpAddr::V4(_) => 4,
+                IpAddr::V6(_) => 16,
+            },
+            Link(_) | FwMark(_) | FlowInfo(_) | Flags(_) => 4,
+            EncapType(_)
+            | EncapFlags(_)
+            | EncapSPort(_)
+            | EncapDPort(_)
+            | Ipv6RdPrefixLen(_)
+            | Ipv6RdRelayPrefixLen(_) => 2,
+            Ttl(_) | Tos(_) | Protocol(_) | PMtuDisc(_) | CollectMetada(_)
+            | EncapLimit(_) => 1,
+            Other(nla) => nla.value_len(),
+        }
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        use self::InfoIpTunnel::*;
+        match self {
+            Ipv6RdPrefix(value) | Ipv6RdRelayPrefix(value) => {
+                buffer.copy_from_slice(&value.octets())
+            }
+            Link(value) | FwMark(value) | FlowInfo(value) => {
+                NativeEndian::write_u32(buffer, *value)
+            }
+            Flags(f) => {
+                NativeEndian::write_u32(buffer, f.bits());
+            }
+            Local(value) | Remote(value) => match value {
+                IpAddr::V4(ipv4) => buffer.copy_from_slice(&ipv4.octets()),
+                IpAddr::V6(ipv6) => buffer.copy_from_slice(&ipv6.octets()),
+            },
+            EncapType(value) => {
+                NativeEndian::write_u16(buffer, (*value).into())
+            }
+            EncapFlags(f) => NativeEndian::write_u16(buffer, f.bits()),
+            EncapSPort(value)
+            | EncapDPort(value)
+            | Ipv6RdPrefixLen(value)
+            | Ipv6RdRelayPrefixLen(value) => {
+                NativeEndian::write_u16(buffer, *value)
+            }
+            Protocol(value) => buffer[0] = i32::from(*value) as u8,
```

Comment:
The `as` is dangerous keyword because it discard data silently.

According to https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml , it is always u8.

So implement `From<u8> for IpProtocol` there instead of using `as`.  When you do that, please fix `src/rule/attribute.rs` also.

Reply:
I implemented `From<u8> for IpProtocol` instead of using `as`. I also fixed `src/rule/attribute.rs`.